server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java

```diff
@@ -616,42 +621,57 @@ public void testComputeConsideringShardSizes() {
 
         var discoveryNodesBuilder = DiscoveryNodes.builder()
             .add(createDiscoveryNode("node-0", DiscoveryNodeRole.roles()))
-            .add(createDiscoveryNode("node-1", DiscoveryNodeRole.roles()));
+            .add(createDiscoveryNode("node-1", DiscoveryNodeRole.roles()))
+            .add(createDiscoveryNode("node-2", DiscoveryNodeRole.roles()));
 
         var metadataBuilder = Metadata.builder();
         var routingTableBuilder = RoutingTable.builder();
 
-        for (int i = 0; i < 10; i++) {
+        var commonIndexSettings = Settings.builder()
+            .put("index.number_of_shards", 1)
+            .put("index.number_of_replicas", 0)
+            .put("index.version.created", Version.CURRENT)
+            .put("index.routing.allocation.exclude._name", "node-2")
+            .build();
+
+        ShardRouting index0PrimaryShard;
+        ShardRouting index0ReplicaShard;
+        {
+            var indexName = "index-0";
+
+            metadataBuilder.put(IndexMetadata.builder(indexName).settings(commonIndexSettings));
+
+            var indexId = metadataBuilder.get(indexName).getIndex();
+            var shardId = new ShardId(indexId, 0);
+
+            index0PrimaryShard = newShardRouting(shardId, "node-1", null, true, STARTED);
```

Comment:
How are we ensuring that this shard is not being relocated to node-0, thus making the replica able to move to node-1, failing the test? I think it works fine, but perhaps you can comment on the assumption and why it holds in a comment in the code.

Reply:
May be it is. During the simulation we are not distinguishing between the primary and the replica. The result balance for this  shard is [node-0, node-1] regardless